---
title: "03 - Alignment"
author: "jerod"
date: "June 29, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```


After looking at our Fastq files, we want to do the actual alignment and quantitation.
We use STAR for alignment and RSEM for quantitation.

Because we're using nonstandard spike-ins, we need to make our own reference files.
```{r}
#connect to sbg:
##my api token. [ should be okay for this 1 HS run]
require(sevenbridges)
sevenbridgesapitoken="68916d51f8ba48c1a87a47f135bafef0"
authurl="https://api.sbgenomics.com/v2/"
a<-Auth(sevenbridgesapitoken,url=authurl)
p<-a$project(name = erccproject)

#aid=p$app(name="SBG FASTQ Merge")
#genomefastalist<-c(p$file(name="SRM2374_ambiguities_resolved_v1.fasta")$id,p$file(name="SIRV_150601a.fasta")$id,p$file(name="human_g1k_v37_decoy.fasta")$id) 
#tsk<-p$task_add(name="Generate Fasta",description="haha",app=aid,inputs=genomefastalist)
#not run - already exists on server.
```

```{r}
#define the correct input files:  fastq, genomefastalist, GTFfile*
gtf="gencodeERCCSIRVO.gtf" 
reffq="prevmerged.fasta" #generated by merging SIRV_150601a.fasta, Homo_Sapiens.Ensembl.GRCh37.fa, and SRM2374_ambiguities_resolves_NoPolyA_v1.FASTA [sbg fastq merge]
gtfid=p$file(name=gtf)$id
refid=p$file(name=reffq)$id

#carefully defined loop:  We want to select files that are:
#1: Not already mapped --- i do not know how to do this part:(
#2: from the same library
#3: appropriately paired.
for(fastq in fastqlist){
  #need a way to differentiate between pairs.  The lazy way that i'm using for now:
  if(grepl("_R1_",fastq)){
  fastq1<-fastq
  next}
  if(grepl("_R2_",fastq)){
  fastq2<-fastq}
  #prepare to run the app 
  if(!is.null(fastq)){
    #temporary because there's a WEIRD bug where the API doesn't find a subset of my files...
    sampleid=p$file(name=fastq)$metadata$sample_id  #this requires that metadata is properly set
    appid<-p$app(name="STAR-RSEM")$id
    
    
    tsk <- p$task_add(name = paste("STAR-rsem",sampleid,sep="_"), 
           description = paste("STAR-rsem on",sampleid,sep="_"), 
           app = appid,
           inputs = list(sample_name=sampleid,reference_fasta_file=p$file(id=refid),gtf=p$file(id=gtfid),input_archive_file=p$file(name=fastq1),input_archive_file_1=p$file(name=fastq2))
           )
#tsk$run() #don't actually run it yet...


}
}
#save the appropriate output files: transcriptome_aligned_reads , reads_per_gene

````