---
title: "03 - Alignment"
author: "jerod"
date: "June 29, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```


After looking at our Fastq files, we want to do the actual alignment and quantitation.
We use STAR for alignment and RSEM for quantitation.

Because we're using nonstandard spike-ins, we need to make our own reference files.
[Approx. Alignment cost:  $2.25*nsample (~$40)]
```{r}
# Repeated housekeeping actions
# connect to sbg:
# Jerod's api token. Change if not jerod!
erccproject="prototypeercc2"
require(sevenbridges)
sevenbridgesapitoken="68916d51f8ba48c1a87a47f135bafef0"
authurl="https://api.sbgenomics.com/v2/"
a<-Auth(sevenbridgesapitoken,url=authurl)
p<-a$project(name = erccproject)
aid=p$app(name="SBG FASTQ Merge")
# Import metadatatable.txt
fastqmetadata <- read.table(file="metadatatable.txt",sep=",",stringsAsFactors = FALSE,header = TRUE,colClasses = c(rep("character",10)))
```
```{r}
genomefastalist<-FilesList(findfile(name="SRM2374_ambiguities_resolved",p),findfile(name="SIRV_150601a.fasta",p),findfile(name="GRCh38_full_analysis_set_plus_decoy",p))
tsk<-p$task_add(name="Generate Fasta",description="ERCC2-Reference-Fasta(SIRV-ERCC-GRCH38)",app=aid$id,inputs=list(fastq=genomefastalist,group_by="Sample"))
tsk$run()
# Creates a merged fasta file for the three -- named 'merged.fastq'
```

```{r}
#define the correct input files:  fastq, genomefastalist, GTFfile*
findfile<-function(name,p,...){
  offset<-0
  while(length(p$file(offset=offset))==100){
    offset<-offset+100
  }# Count the total number of files
  maxoffset<-offset
  offset<-0
  while(offset <= maxoffset){
    if(!is.null(p$file(name=name,offset=offset))){return(p$file(name=name,offset=offset))}
    offset<-offset+100}
return(NULL)
} # findfile 
gtflist <- FileList(findfile(name="ERCC.gtf",p),findfile(name="SIRV_O_150601a.gtf",p),findfile(name="Homo_sapiens.GRCh38.84.gtf",p))
gtfcombine <- p$task_add(app=p$app(name="GTF combiner")$id,inputs=list(Infiles=FilesList(findfile(name="ERCC.gtf",p),findfile(name="SIRV_O_150601a.gtf",p),findfile(name="Homo_sapiens.GRCh38.84.gtf",p))))
gtfcombine$run()

reffq="merged.fastq" # generated by merging SIRV_150601a.fasta, Homo_Sapiens.GRCh38.84.fa, and SRM2374_ambiguities_resolves_NoPolyA_v1.FASTA [sbg fastq merge]
refgtf="_1_merged.gtf" # until it's renamed, at least.  

sbfastqlist<-lapply(fastqmetadata$Fastq.filename,FUN = function(x){findfile(name=x,p=p)})
tsklist<-NULL
# PRODUCTION VERSION
# appid<-p$app(name="RSEM Prepare Reference")$id
# ntsk<-p$task_add(name = "GenerateReference",app = appid,inputs = list(reference_name="STAR_RSEM_GRCh38_ERCC_SIRV",bowtie2=FALSE,star=TRUE,bowtie=FALSE,star_sjdboverhang=100,
#                                                                     reference_fasta_file=findfile(name = reffq,p = p),gtf=findfile(refgtf,p=p)))
# ntsk$run()

# TESTING VERSION
 ntsk<-p$task(id="0c0239f9-b871-40f0-b9f6-4f0de93cf6ec") # result of the one run i did: using above code.
# task to prepare reference
referencelist=ntsk$file() 

# carefully defined loop:  We want to select files that are:
# 1: TODO:  Not already mapped 
# 2: from the same library
# 3: appropriately paired.
for(fastq in sbfastqlist){
  
  # metadata is set now:  file$meta()$paired_end=="2" #yes, it's still a string
  switch(fastq$meta()$paired_end,
"1" = {fastq1<-fastq;next},
"2" = {fastq2<-fastq})
# TODO check to see if the output exists.
# check to make sure that fastq1 and fastq2 are actually the same sample
  if(fastq1$meta()$batch_number!=fastq2$meta()$batch_number){stop("fastq1 and fastq2 are not the same sample: please sort your fastqmetadata file by sample!")}
# prepare to run the app 
  
    sampleid=fastq1$metadata$batch_number
    if(is.null(sampleid)){stop("batch_number not set for ",fastq)}
    appid<-p$app(name="STAR-RSEM")$id
    
    tsk <- p$task_add(name = paste("STAR-rsem",sampleid,sep="_"), 
           description = paste("STAR-rsem on",sampleid,sep="_"), 
           app = appid,
           inputs = list(sample_name=sampleid,prepare_reference_output_files=referencelist,input_archive_file=fastq1,input_archive_file_1=fastq2)
           )
    tsklist<-c(tsklist,tsk$id) # save a list of the tasks that're associated with this phase, to ease the process of collecting the output files somewhat.

    tsk$run() # runs the alignment/quant.
}
# TESTING 
print(tsklist)

# TODO identify the appropriate output files for the next steps: transcriptome_aligned_reads, reads_per_gene, bam
# TODO generate app and task to merge all reads_per_isoform files to make the master count table [05 - count table.rmd]
# TODO 
````