---
title: "03 - Alignment"
author: "jerod"
date: "June 29, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```


After looking at our Fastq files, we want to do the actual alignment and quantitation.
We use STAR for alignment and RSEM for quantitation.

Because we're using nonstandard spike-ins, we need to make our own reference files.
[Approx. Alignment cost:  $2.25*nsample (~$40)]
```{r}
#connect to sbg:
##my api token. [ should be okay for this 1 HS run]
require(sevenbridges)
sevenbridgesapitoken="68916d51f8ba48c1a87a47f135bafef0"
authurl="https://api.sbgenomics.com/v2/"
a<-Auth(sevenbridgesapitoken,url=authurl)
p<-a$project(name = erccproject)

#aid=p$app(name="SBG FASTQ Merge")
#genomefastalist<-c(p$file(name="SRM2374_ambiguities_resolved_v1.fasta")$id,p$file(name="SIRV_150601a.fasta")$id,p$file(name="human_g1k_v37_decoy.fasta")$id) 
#tsk<-p$task_add(name="Generate Fasta",description="haha",app=aid,inputs=genomefastalist)
#not run - already exists on server.
```

```{r}
#define the correct input files:  fastq, genomefastalist, GTFfile*

# TODO Switch to GRCh38 when it ships to collaborators... SM

gtf="gencodeERCCSIRVO.gtf" 
reffq="merged.fasta" #generated by merging SIRV_150601a.fasta, Homo_Sapiens.Ensembl.GRCh37.fa, and SRM2374_ambiguities_resolves_NoPolyA_v1.FASTA [sbg fastq merge]
#gtfid=p$file(name=gtf)$id
#refid=p$file(name=reffq)$id

# Runonce
appid<-p$app(name="RSEM Prepare Reference")$id
ntsk<-p$task_add(name = "GenerateReference",app = appid,inputs = list(reference_name="STAR_RSEM_HG_ERCC_SIRV",bowtie2=FALSE,star=TRUE,bowtie=FALSE,star_sjdboverhang=100,
                                                                     reference_fasta_file=findfile(name = "merged.fasta",p = p),gtf=findfile("gencodeERCCSIRVO.gtf",p=p)))
ntsk$run()
# task to prepare reference
referencelist=ntsk$file() 

#carefully defined loop:  We want to select files that are:
#1: Not already mapped --- i do not know how to do this part:(
#2: from the same library
#3: appropriately paired.
for(fastq in fastqlist){
  
  #metadata is set now:  file$meta()$paired_end=="2" #yes, it's still a string
  if(findfile(name=fastq,p)$metadata$paired_end=="1"{
  fastq1<-fastq
  next}
  if(findfile(name=fastq,p)$metadata$paired_end=="2"{
  fastq2<-fastq}
  #prepare to run the app 

    sampleid=findfile(name=fastq,p)$metadata$batch_number
    if(is.null(sampleid)){stop("batch_number not set for ",fastq)}
    appid<-p$app(name="STAR-RSEM")$id
    
    tsk <- p$task_add(name = paste("STAR-rsem",sampleid,sep="_"), 
           description = paste("STAR-rsem on",sampleid,sep="_"), 
           app = appid,
           inputs = list(sample_name=sampleid,prepare_reference_output_files=referencelist,input_archive_file=p$file(name=fastq1),input_archive_file_1=p$file(name=fastq2))
           )
    tsklist<-c(tsklist,tsk$id) # save a list of the tasks that're associated with this phase, to ease the process of collecting the output files somewhat.

    tsk$run() #runs the alignment/quant.
}

# TODO identify the appropriate output files for the next steps: transcriptome_aligned_reads, reads_per_gene, bam
# TODO generate app and task to merge all reads_per_isoform files to make the master count table [05 - count table.rmd]
# TODO 
````