---
title: "04 - BAM QC"
author: "jerod"
date: "June 29, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```

This is a chance to run some of the read-level QC tools & view their outputs:

1 - Picard RNASeqMetrics 
```{r}
# generate a flat reference for picard...
#app<-p$app(name="makeflatref")$id
#gfr<-p$task_add(app=app,inputs=list())
#for(thebam in bamlist){
#app<-p$app(name="Picard CollectRnaSeqMetrics")$id
#RSM<-p$task_add(app=p$app(name="GTF combiner")$id,inputs=list(input_bam=findfile(name=thebam,p),reference=fasta,ref_flat=flatref)))
# Picard requires a weirdly-formatted text file as an input.  Ones for the genome can be found on UCSC, but controls??
# Aborting this for now
#}
```
2 - RNASEQC
* Coverage Plots
* GC Content plot

```{r}
# Re-define some previous data...
# TODO: Merge all of the RMDs together where appropriate.

reffq="merged.fastq" # generated by merging SIRV_150601a.fasta, Homo_Sapiens.GRCh38.84.fa, and SRM2374_ambiguities_resolves_NoPolyA_v1.FASTA [sbg fastq merge]
refgtf="merged.gtf" # until it's renamed, at least.  

# Preliminary:  Index the fastq [and maybe rename it to a fasta.]
findfile(reffq)$copyTo(project = "lharris/ercc2pipeline",name="merged.fasta")
ifq<-p$task_add(app=p$app(name="SAMtools Index FASTA")$id,inputs=list(input_fasta_file=findfile("merged.fasta"))) # TODO: Replace all of the instances of 'merged.x' with something that's more appropriate.


# Run RNA-sEQC
bamlist<-p$file(name=".sorted.edited.bam$",limit=0) # Temporary easy identification of files for testing purposes. 
# TODO: Make this search tsklist or something.
RSEQC<-p$task_add(app=p$app(name="RNA-SeQC")$id,inputs=list(bam_files=bamlist,transcripts=findfile(refgtf,exact=TRUE),reference=findfile("merged.fasta",exact=TRUE),transcript_type=2))
# TODO: Make a GTF that RSEQC actually likes [it's not a fan of the GENE entries in the gtf: i guess i'll just do a grep -v ?]
# TODO: Figure out what ACTUALLY is required to make a good gtf, because that is insufficient...
# TODO: If the output proves relevant, add transcript_details=TRUE and strictMode=TRUE)
```