---
title: "05 - Count Tables"
author: "jerod"
date: "June 29, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```

```{r setup2, echo=FALSE}
require(sevenbridges)
sevenbridgesapitoken="68916d51f8ba48c1a87a47f135bafef0"
authurl="https://api.sbgenomics.com/v2/"
a<-Auth(sevenbridgesapitoken,url=authurl)
p<-a$project(name = "prototypeercc2",exact=TRUE)
source("ConvenienceFunctions.R") # Load convenience functions such as gmeta
```

```{r}
metadatatable<-read.csv("metadatatable.txt")
Labid<-levels(as.factor(metadatatable$investigation))
# merge all of the individual run counts into a single master count table
sampleglist<-p$file(name="genes.results",limit=0) # this is specific to the STAR/RSEM output
sampletlist<-p$file(name="isoforms.results",limit=0) # as is this.
# TODO: Make the generation of sampleglist and sampletlist not specific to particular outputs [use tkslist? A switch statement?]

# this app is defined by dmergecounts.R and runs mergecounts.R
gnames<-lapply(sampleglist,FUN = gmeta) # extracts the metadata
gnames<-unlist(lapply(gnames,"[[","sample_name")) # extracts just the samplenames
tnames<-lapply(sampletlist,FUN = gmeta)
tnames<-unlist(lapply(tnames,"[[","sample_name"))

# minor sanity checking:
if(length(sampleglist)!=length(gnames)){stop("Number of sample names(glist) is not equal to number of samples found.")}
if(length(sampletlist)!=length(tnames)){stop("Number of sample names(tlist) is not equal to number of samples found.")}

gtask<-p$task_add(app=p$app(name="mergecounts")$id,inputs=list(source_file=sampleglist,outputfilename=paste0(Labid,"_Gene_Counts"),Column=5,names=paste(gnames,collapse=","))) # Column 5 is the expected count value.  Again, specific to star/rsem.
ttask<-p$task_add(app=p$app(name="mergecounts")$id,inputs=list(source_file=sampletlist,outputfilename=paste0(Labid,"_Transcript_Counts"),Column=5,names=paste(tnames,collapse=","))) 
gtask$run()
ttask$run()
```

