---
title: "07 - DE Testing"
author: "jerod"
date: "June 29, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache = TRUE)#have to cache for some of this DE to be iteratively useful.
source("ConvenienceFunctions.R")
requirelist(bcpackages=c("S4Vectors","GenomicFeatures","SummarizedExperiment"),ghpackages = "jeparson/mixturesolutions",crpackages = c("ggplot2","plotly")
experimentaldesignfile="SampleDesign.csv"
```

Now to calculate differential expression.

We need to make a useful R object with all of the experimental data in it.
```{r,echo=TRUE}
# The counts
mastertable<-"tools/tests/NIST_MM1_Transcript_Counts.csv"
mastercounts<-read_csv(mastertable)
# ForNow: Let's make these counts integer.
mastercounts[,2:length(mastercounts)]<-round(mastercounts[,2:length(mastercounts)])
expdesign<-S4Vectors::DataFrame(read.table(experimentaldesignfile,sep="\t",header=TRUE))
# not sure why i made it an s4df.
mastercounts<-mastercounts[,c("transcript_id",as.character(expdesign$sample))] # These seem to get merged out of order, so a quick fix.
# some minor expectation that this won't work in production, but getting the order correct is important.

# The metadata 
```
Just to double check that there are no significant amounts of SIRVs present in the samples
where we didn't add SIRVs:
```{r sanity}
# Expectation NOSIRV: S11/S18/S5/S12/S6/S17: Perfect 
colSums(mastercounts[grep("SIRV",mastercounts$transcript_id),2:length(mastercounts)])>(colSums(mastercounts[,2:length(mastercounts)])*0.01)
```

We then need to do DE testing of various pairwise comparisons.

```{r,echo=FALSE}
dds<-DESeqDataSetFromMatrix(mastercounts[,2:length(mastercounts)],colData = expdesign[1:18,],design = ~ mix + erccspike + sirvspike + sirvspike:mix) # the interaction term being the relevant measure of interference
dds<-DESeq(dds)
rownames(dds)<-mastercounts[,1]
```
A heatmap of the samples should make a checkerboard of the 3 mixes and be subclustered on the 3 different SIRV mixes.
```{r}

library("pheatmap")
library("RColorBrewer")
rld <- rlog(dds, blind=FALSE)
sampleDists <- dist( t( assay(rld) ) )
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste(colData(rld)$mix,colData(rld)$sirvspike,colData(rld)$erccspike)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
plotPCA(rld,intgroup=c("mix","sirvspike"))
```
```{r}
##TODO:  replace all of the quoted things here with the appropriate references to the expdesign file.
res<-results(dds,contrast = c("mix","M","L")) # There should be a great deal of DE
res2<-results(dds,contrast=c("erccspike","Mix1","Mix2")) # There should be very little..
res3<-results(dds,name="sirvspike_E2_vs_E1") # There should be very little..
res4<-results(dds,contrast=c("sirvspike","E1","none")) # There should be very little..
res5<-results(dds,contrast=c("sirvspike","E2","none")) # There should be very little..
res6<-results(dds,name="mixM.sirvspikeE2")# There should be none..
res7<-results(dds,name="mixLM.sirvspikeE2")# There should be none..
res8<-results(dds,name="mixLM.sirvspikenone")# There should be none..

## And now to do fun things to collect the appropriate demonstrations....
# Table of Number of padjs <0.05 outside of the 'expected' range/value... 

# Collect a few useful vectors
spikepattern<-"^SIRV|^DQ|^EN[0-9]" # TODO: update for sequins + pombe once the correct reference exists 
isercc<-grep("^DQ|EN",mastercounts$transcript_id)
issirv<-grep("SIRV",mastercounts$transcript_id)
isspike<-c(issirv,isercc)
spikevec<-c(rep(0,length(rld)))
spikevec[isspike]<-1
theme_set(theme_bw(base_size=18))
a<-ggplot(data=as.data.frame(res2))+geom_point(aes(x=log2(baseMean),y=log2FoldChange,col=padj<0.05,size=as.numeric(1+spikevec),pch=as.factor((spikevec)))) # Easy MA plots.  TODO: Annotate with SIRV/ERCC status so they can pop the way they're supposed to.  Size? Alpha? Col?
plotly::ggplotly(a)
```
The DE results should show that there are no DEG between samples that aren't different [besides the spikes]
```{r}
test<-res4[which(res4$padj<0.05),] #other than SIRVs, nothing should be DE 
hist(test[grep("SIRV|^DQ|^EN[0-9]",rownames(test),invert=TRUE),]$log2FoldChange) # is there a bias in the direction of fold changes?  If SIRVs were interfering...

```

```{R}
# Variance vs Concentration [spikes vs endo]

# Coefficient of Deviation [spikes vs endo]

# Binned apparent LogFoldChanges for endogenous transcripts vs Spike-ins

# RMSD from mixture model is unchanged:

mixobj<-mixturesolutions::mixmetrics(mrnatype = "ercc",prenormalized = TRUE,modeltype="threemix",componentnames=c("LM","M","L"),mixnames = c("LM","M","L"),indf = mastercount,dds=expdesign,trueproportions = data.frame(L=c(2,-1,0),M=c(2,0,-1),LM=c(0,.5,.5)))
# This is extraordinarily incomplete.  MixMetrics doesn't handle threemix at all, and it tries to automatically do collapsereplicates which is inappropriate here, among other things.
# However, magicmodelsolve more or less says what it needs to do, even if it doesn't do it yet.

# 
```
And finally just for posterity...
```{r}
sessionInfo()
```